File: javascript/glassdoor.js
"use strict";

const webdriver = require('selenium-webdriver');
const browser = new webdriver.Builder().usingServer().withCapabilities({
  'browserName': 'chrome'
}).build();

function clickLink(link) {
  link.click();
}

function handleFailure(err) {
  console.error('Something went wrong\n', err.stack, '\n');
}

function closeBrowser() {
  browser.quit();
}

function findJobs() {
  // 'sc.keyword' is the name of search input field
  browser.findElement(webdriver.By.id('KeywordSearch')).sendKeys(
    'software engineer'
  );

  // 'LocationSearch' is the id of location input field
  let location = browser.findElement(webdriver.By.id('LocationSearch'));
  // clears input field of default values before sending keys
  location.clear();
  location.sendKeys(
    'San Francisco, CA'
  );

  // finds search button and clicks
  return browser.findElement(webdriver.By.className('search'));
}

function clickJobs() {
  // Targets only the job listings that allow you to apply through glassdoor.
  // 'applyText' is the class name of the green text that says 'Apply on Glassdoor'
  browser.findElements(webdriver.By.className("applyText")).then(elements => {
    for (let i = 0; i < elements.length; i++) {
      // Targets the first glassdoor application job listing.
      if (i === 0) {
        let element = elements[i];
        element.click().then(function() {
          // closes modal that pops up after clicking the first job listing
          browser.findElement(webdriver.By.className('mfp-close')).click()
          .then(() => {
            setTimeout(() => {
              // finds and clicks the 'apply' button
              let applyButton = browser.findElement(
                // Seems that glassdoor changes the name of their apply button
                // twice a day, switching from the href to class. Keep this
                // commented out, and we can switch them whenever the
                // attributes change
                webdriver.By.css('[href="#easy-apply-module"]')
                // webdriver.By.className('ezApplyBtn')
                // webdriver.By.className('applyButton')
              );

              // Sometimes, Selenium wouldn't be able to locate the
              // 'Apply on Glassdoor' button and would throw a
              // 'ElementNotVisibleError'. I suspected it had to do with
              // async requests. After hours, I found a solution that works:
              // The following code waits until applyButton is found, then
              // users the browser's methods and moves the mouse above the
              // applyButton element, then clicks it. It hasn't broken on
              // me yet, and if it does, Idk how to fix.
              applyButton.then(() => {
                browser.actions().mouseMove(applyButton).click().perform();
              });
            }, 5000);

          });
        });
      }
    }
  });
}

function apply() {
  // Fills out 'name' input field in application form
  browser.findElement(webdriver.By.id('ApplicantName')).sendKeys(
    'John Doe'
  );

  // Fills out 'email' input field in application form
  browser.findElement(webdriver.By.id('ApplicantEmail')).sendKeys(
    'john_doe@gmail.com'
  );

  // Fills out the optional 'message' field
  browser.findElement(webdriver.By.name('coverLetterHTML')).sendKeys(
    'This is my cover letter'
  );

}

browser.get('https://www.glassdoor.com/index.htm');
// Using setTimeout to prevent spamming requests to prevent the server
// from blocking the webcrawler. Ideally, clickJobs should be invoked
// after findJobs, but I ran into a few issues have yet to figure out
//
// Ideally, we would want something like this:
  // browser.wait(findJobs, 5000).then(clickJobs)
  // But this throws errors
// setTimeout(findJobs, 5000);
// setTimeout(clickJobs, 25000);
// setTimeout(apply, 35000);

setTimeout(() => (
  findJobs().click().then(() => (
    setTimeout(() => {
      clickJobs();
      setTimeout(apply, 10000);
    }, 5000)
  ))
), 5000);

File: ruby/glassdoor.rb
require_relative "../modules/selenium.rb"
require_relative "../modules/io.rb"
require "byebug"

include IOStream
include Selenium

DRIVER = Selenium::DRIVER
WAIT = Selenium::WebDriver::Wait.new(timeout: 60)

def handle_failure(err)
  puts "Something went wrong\n"
end

def closeBrowser
  DRIVER.quit
end

def glassdoor_signin
  signin
end

def find_jobs
  location = IOStream::input_locations
  position = IOStream::input_positions

  location_input = DRIVER.find_element(id: "LocationSearch")
  location_input.clear
  location_input.send_keys("San Francisco, CA")

  position_input = DRIVER.find_element(id: "KeywordSearch")
  position_input.send_keys(position)

  DRIVER.find_element(class: "search")
end

def get_job_info
  # glassdoor_job_listings selects only the listings where you can apply
  # directly through glassdoor
  glassdoor_job_listings = DRIVER.find_elements(class: 'applyText')

  # job_listings = DRIVER.find_elements(class: 'jobListing')
  companies= DRIVER.find_elements(css: 'span.showHH.inline.empName')
  locations = DRIVER.find_elements(css: 'span.small.location')
  titles = DRIVER.find_elements(css: 'span.title')

  glassdoor_job_listings.each.with_index do |listing, i|
    break if i > 0

    listing.click

    sleep(1)

    DRIVER.find_element(class: 'mfp-close').click if i == 0

    description = DRIVER.find_elements(
      css: 'div.jobDescriptionContent.desc'
    )

    sleep(1)

    DRIVER.find_element(class: 'ezApplyBtn').click

    sleep(1)
    apply
  end

end

def apply
  DRIVER.find_element(id: 'ApplicantName').send_keys('John Doe')
  sleep(1)
  DRIVER.find_element(id: 'ApplicantEmail').send_keys('john_doe@gmail.com')
  sleep(1)
  DRIVER.find_element(name: 'coverLetterHTML').send_keys(
    'This is my cover letter.'
  )
  sleep(1)
  gmail_signin

  DRIVER.switch_to.window DRIVER.window_handles.first

byebug

  resume_select = DRIVER.find_element(id: "ExistingResumeSelectBoxIt")
  resume_select.click
  sleep(1)
  resume_file = DRIVER.find_element(data_val: "resume.pdf")
  resume_file.click

  byebug
end

def gmail_signin
  glassdoor_window = DRIVER.window_handle

  DRIVER.find_element(id: 'GoogleDriveBtn').click

  sleep(1) until DRIVER.window_handles.count == 2

  DRIVER.switch_to.window DRIVER.window_handles.last

  email_input = DRIVER.find_element(id: 'Email')
  email_input.send_keys('')
  next_button = DRIVER.find_element(id: 'next')
  next_button.click
  sleep(1)
  password_input = DRIVER.find_element(id: 'Passwd')
  password_input.send_keys('')
  signin_button = DRIVER.find_element(id: 'signIn')
  signin_button.click
end

DRIVER.get("https://www.glassdoor.com/index.htm")

sleep(10)

sleep(1)
find_jobs.click
sleep(1)
get_job_info
sleep(90000)
